acme/dns_challenge.go

```diff
@@ -62,48 +63,125 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
+	if err != nil {
+		return false, fmt.Errorf("Could not lookup nameservers: %v", err)
+	}
+
+	for _, ns := range nameservers {
+		r, err := dnsQuery(fqdn, dns.TypeTXT, net.JoinHostPort(ns, "53"))
+		if err != nil {
+			return false, err
+		}
+
+		if r.Rcode != dns.RcodeSuccess {
+			return false, fmt.Errorf("NS %s returned RCode %s", ns, dns.RcodeToString[r.Rcode])
+		}
+
+		var found bool
+		for _, rr := range r.Answer {
+			if _, ok := rr.(*dns.TXT); ok {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false, fmt.Errorf("NS %s did not return expected TXT record", ns)
+		}
+	}
+
+	return true, nil
+}
+
+// dnsQuery directly queries the given authoritative nameserver.
+func dnsQuery(fqdn string, rtype uint16, nameserver string) (in *dns.Msg, err error) {
 	m := new(dns.Msg)
-	m.SetQuestion(domain+".", dns.TypeSOA)
-	c := new(dns.Client)
-	in, _, err := c.Exchange(m, "google-public-dns-a.google.com:53")
+	m.SetQuestion(fqdn, rtype)
+	m.SetEdns0(4096, false)
+	m.RecursionDesired = false
+
+	in, err = dns.Exchange(m, nameserver)
+	if err == dns.ErrTruncated {
+		tcp := &dns.Client{Net: "tcp"}
+		in, _, err = tcp.Exchange(m, nameserver)
+	}
+
+	return
+}
+
+// lookupNameservers returns the authoritative nameservers for the given domain name.
+func lookupNameservers(fqdn string) ([]string, error) {
+	var referralNameservers []string
+
+	// We start recursion at the gTLD origin
+	// so we don't have to manage root hints
+	labels := dns.SplitDomainName(fqdn)
+	tld := labels[len(labels)-1]
+	nss, err := net.LookupNS(tld)
 	if err != nil {
-		return false
+		return nil, fmt.Errorf("Failed to resolve TLD %s %v", tld, err)
 	}
 
-	var authorativeNS string
-	for _, answ := range in.Answer {
-		soa := answ.(*dns.SOA)
-		authorativeNS = soa.Ns
+	for _, ns := range nss {
+		referralNameservers = append(referralNameservers, ns.Host)
 	}
 
-	fallbackCnt := 0
-	for fallbackCnt < preCheckDNSFallbackCount {
-		m.SetQuestion(fqdn, dns.TypeTXT)
-		in, _, err = c.Exchange(m, authorativeNS+":53")
-		if err != nil {
-			return false
+	// Follow the referrals until we hit the authoritative NS
+	for depth := 0; depth < recursionMaxDepth; depth++ {
+		var r *dns.Msg
+		var err error
+		
+		for _, ns := range referralNameservers {
+			r, err = dnsQuery(fqdn, dns.TypeNS, net.JoinHostPort(ns, "53"))
+			if err != nil {
+				continue
+			}
+			if r.Rcode == dns.RcodeSuccess {
+				break
+			}
+			if r.Rcode == dns.RcodeNameError {
```

Comment:
Hi Willy! Thank your for reviewing the code! :smiley: 

> This ultimately causes lego to spin forever

No it won't. Lego will gracefully exit after 60 seconds, which should be plenty of time for the user  to think about why he is trying to issue a certificate for a nonexistent domain name.

While it might be technically legal to do that under the ACME spec, i don't see why we should specifically enable this behavior. In fact none of the other challenge methods support this scenario.

This being said, i do agree that the very purpose of the `checkDnsPropagation` func is to verify the validation domain rather than the certificate domain. Hence, we should pass the fqdn in the `lookupNameservers` func, which incidentally should solve the usage scenario you described.